external/source/HostingCLR_inject/HostingCLR/HostingCLR.cpp

```diff
@@ -56,119 +59,132 @@ int executeSharp(LPVOID lpPayload)
 	BOOL bLoadable;
 	ICorRuntimeHost* pRuntimeHost = NULL;
 	IUnknownPtr pAppDomainThunk = NULL;
-	_AppDomainPtr pDefaultAppDomain = NULL;
+	_AppDomainPtr pCustomAppDomain = NULL;
 	_AssemblyPtr pAssembly = NULL;
 	SAFEARRAYBOUND rgsabound[1];
-	SIZE_T readed;
 	_MethodInfoPtr pMethodInfo = NULL;
 	VARIANT retVal;
 	VARIANT obj;
-	SAFEARRAY *psaStaticMethodArgs;
+	SAFEARRAY* psaStaticMethodArgs;
 	VARIANT vtPsa;
 
-	unsigned char pSize[8];
+	Metadata metadata;
 
-	//Read parameters assemblysize + argssize
-	ReadProcessMemory(GetCurrentProcess(), lpPayload, pSize, 8, &readed);
+	// Structure of lpPayload:
+	// - Packed metadata, including lengths of the following fields
+	// - Pipe name (ASCII)
+	// - Appdomain name (ASCII)
+	// - Clr Version Name (ASCII)
+	// - Param data
+	// - Assembly data
+	
+	memcpy(&metadata, lpPayload, METADATA_SIZE);
 
-	PARAMSIZE assemblysize;
-	assemblysize.myByte[0] = pSize[0];
-	assemblysize.myByte[1] = pSize[1];
-	assemblysize.myByte[2] = pSize[2];
-	assemblysize.myByte[3] = pSize[3];
+	BYTE* data_ptr = (BYTE*)lpPayload + METADATA_SIZE;
 
-	PARAMSIZE argssize;
-	argssize.myByte[0] = pSize[4];
-	argssize.myByte[1] = pSize[5];
-	argssize.myByte[2] = pSize[6];
-	argssize.myByte[3] = pSize[7];
+	char* pipeName = (char*)malloc((metadata.pipenameLength + 1) * sizeof(char));
+	memcpy(pipeName, data_ptr, metadata.pipenameLength);
+	pipeName[metadata.pipenameLength] = 0; // Null-terminate
+	data_ptr += metadata.pipenameLength;
 
-	long raw_assembly_length = assemblysize.intvalue;
-	long raw_args_length = argssize.intvalue;
+	char* appdomainName = (char*)malloc((metadata.appdomainLength + 1) * sizeof(char));
+	memcpy(appdomainName, data_ptr, metadata.appdomainLength);
+	appdomainName[metadata.appdomainLength] = 0; // Null-terminate
+	data_ptr += metadata.appdomainLength;
 
-	unsigned char *allData = (unsigned char*)malloc(raw_assembly_length * sizeof(unsigned char)+ raw_args_length * sizeof(unsigned char) + 9 * sizeof(unsigned char));
-	unsigned char *arg_s = (unsigned char*)malloc(raw_args_length * sizeof(unsigned char));
-	unsigned char *rawData = (unsigned char*)malloc(raw_assembly_length * sizeof(unsigned char));
+	char* clrVersion = (char*)malloc((metadata.clrVersionLength + 1) * sizeof(char));
+	memcpy(clrVersion, data_ptr, metadata.clrVersionLength);
+	clrVersion[metadata.clrVersionLength] = 0; // Null-terminate
+	data_ptr += metadata.clrVersionLength;
 
-	SecureZeroMemory(allData, raw_assembly_length * sizeof(unsigned char) + raw_args_length * sizeof(unsigned char) + 9 * sizeof(unsigned char));
-	SecureZeroMemory(arg_s, raw_args_length * sizeof(unsigned char));
-	SecureZeroMemory(rawData, raw_assembly_length * sizeof(unsigned char));
+	// Convert to wchar
+	wchar_t* clrVersion_w = new wchar_t[metadata.clrVersionLength + 1];
+	mbstowcs(clrVersion_w, clrVersion, metadata.clrVersionLength + 1);
+	
+	BYTE* arg_s = (unsigned char*)malloc(metadata.argsSize * sizeof(BYTE));;
```

Comment:
```suggestion
	BYTE* arg_s = (BYTE*)malloc(metadata.argsSize * sizeof(BYTE));
```

This also needs to be freed.